sycl/test-e2e/Matrix/common.hpp

```diff
@@ -0,0 +1,82 @@
+#include <random>
+#include <sycl/sycl.hpp>
+
+using bfloat16 = sycl::ext::oneapi::bfloat16;
+
+constexpr float BF16_EPSILON = 0.00781250;
+
+template <typename T, size_t NUM_ROWS, size_t NUM_COLS> struct big_matrix {
+public:
+  T *mat;
+
+public:
+  T *get_data() { return mat; }
+  void set_data(T *data) { mat = data; }
+  big_matrix(T *data) : mat(data) {}
+};
+
+float make_fp32(bfloat16 x) {
+  unsigned int y = *((int *)&x);
+  y = y << 16;
+  float *res = reinterpret_cast<float *>(&y);
+  return *res;
+}
+
+void matrix_multiply_ref(bfloat16 *A, bfloat16 *B, float *C, int MATRIX_M,
+                         int MATRIX_N, int MATRIX_K) {
+  for (unsigned int i = 0; i < MATRIX_M; i++) {
+    for (unsigned int k = 0; k < MATRIX_K; k++) {
+      for (unsigned int j = 0; j < MATRIX_N; j++) {
+        C[i * MATRIX_N + j] +=
+            make_fp32(A[i * MATRIX_K + k]) * make_fp32(B[k * MATRIX_N + j]);
+      }
+    }
+  }
+}
+
+template <typename T>
+void matrix_vnni(unsigned int rows, unsigned int cols, T *src, T *dest,
+                 unsigned int vnniFactor = 2) {
+  for (unsigned int i = 0; i < rows / vnniFactor; i++) {
+    for (unsigned int j = 0; j < cols; j++) {
+      for (unsigned int k = 0; k < vnniFactor; k++) {
+        dest[i * cols * vnniFactor + j * vnniFactor + k] =
+            src[(i * vnniFactor + k) * cols + j];
+      }
+    }
+  }
+}
+
+template <typename T>
+void matrix_fill(unsigned int rows, unsigned int cols, T *src, T val) {
+  for (unsigned int i = 0; i < rows; i++) {
+    for (unsigned int j = 0; j < cols; j++) {
+      src[i * cols + j] = val;
+    }
+  }
+}
+
+template <typename T>
+void matrix_rand(unsigned int rows, unsigned int cols, T *src) {
+  std::random_device dev;
+  std::uniform_real_distribution<float> fdistr(-1.0, 1.0);
+
+  for (unsigned int i = 0; i < rows; i++) {
+    for (unsigned int j = 0; j < cols; j++) {
+      src[i * cols + j] = T(fdistr(dev));
+    }
+  }
+}
+
+template <typename T1, typename T2>
+bool matrix_compare(unsigned int rows, unsigned int cols, T1 *src, T2 *ref) {
+  bool res = true;
+  for (int i = 0; i < rows; i++) {
+    for (int j = 0; j < cols; j++) {
+      if ((fabs(src[i * cols + j] - (T2)ref[i * cols + j])) > BF16_EPSILON) {
```

Comment:
ref is T2*, should we convert to T1 instead? I'm also wondering how the test passes with this problem?
```suggestion
      if ((fabs(src[i * cols + j] - (T1)ref[i * cols + j])) > BF16_EPSILON) {
```

Reply:
Thanks for catching this. 
The test is passing as the conversion is insensitive to values between -1 and +1. meaning the 2 first bits in the mantissa were identical, the third was still under the EPSILON we are using.
I will change the initialization value to something that makes the test fail if we skip the cast.